src/main/kotlin/org/rust/ide/refactoring/RsImportOptimizer.kt

```diff
@@ -71,14 +76,29 @@ class RsImportOptimizer : ImportOptimizer {
             return true
         }
 
+        private fun removeUseSpeck(useSpeck: RsUseSpeck) {
+            if (useSpeck.parent is RsUseGroup) {
+                val nextComma = useSpeck.getNextNonCommentSibling()?.takeIf { it.elementType == RsElementTypes.COMMA }
+                val prevComma = useSpeck.getPrevNonCommentSibling()?.takeIf { it.elementType == RsElementTypes.COMMA }
+                (nextComma ?: prevComma)?.delete()
+                useSpeck.delete()
+            }
+            // else can't delete useSpeck.parent if it is RsUseItem, because it will cause invalidation exception
```

Comment:
I hit this error when I was trying to create a quick fix to remove an (unused) `RsUseItem`. When I removed `RsUseItem`, name resolution failed because it was caching it somehow probably. Is it possible to remove a whole `RsUseItem` currently?

Reply:
Here invalidation exception will happen just because we will [use that `RsUseItem` later](https://github.com/intellij-rust/intellij-rust/blob/029836c47f3c0710f6a40bf68ad0e44edf9dabb3/src/main/kotlin/org/rust/ide/refactoring/RsImportOptimizer.kt#L104-L107). In general I think there should be no problems when removing `RsUseItem`s. Could you share quick fix code where exception happens?

Reply:
Sorry, it confused me, it's a different error, but happens in a similar situation. Try [this](https://github.com/Kobzol/intellij-rust/tree/quick-fix-remove-use-speck) branch, run CLion and remove an unused import:
```rust
mod foo {
    pub struct S;
}

use foo::S;/*caret*/
```

It triggers:
```
In file: file:///projects/personal/intellij-rust/exampleProject/src/main.rs

java.lang.IllegalStateException: Element outside of module: foo
	at org.rust.lang.core.psi.ext.RsStubbedElementImpl.getContainingMod(RsElement.kt:114)
	at org.rust.lang.core.psi.ext.RsPathImplMixin.getContainingMod(RsPath.kt:164)
	at org.rust.lang.core.resolve.NameResolutionKt.processUnqualifiedPathResolveVariants(NameResolution.kt:512)
	at org.rust.lang.core.resolve.NameResolutionKt.processPathResolveVariants(NameResolution.kt:356)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePath$result$1.invoke(RsPathReferenceImpl.kt:144)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePath$result$1.invoke(RsPathReferenceImpl.kt)
	at org.rust.lang.core.resolve.ProcessorsKt.collectPathResolveVariants(Processors.kt:101)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath(RsPathReferenceImpl.kt:143)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath$default(RsPathReferenceImpl.kt:142)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:131)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:129)
	at org.rust.lang.core.resolve.ref.RsResolveCache$resolveWithCaching$$inlined$run$lambda$1.compute(RsResolveCache.kt:99)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:42)
	at org.rust.lang.core.resolve.ref.RsResolveCache.resolveWithCaching(RsResolveCache.kt:99)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedCachedMultiResolve(RsPathReferenceImpl.kt:89)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedMultiResolve(RsPathReferenceImpl.kt:85)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedResolve(RsPathReferenceImpl.kt:74)
	at org.rust.lang.core.resolve.NameResolutionKt.processQualifiedPathResolveVariants(NameResolution.kt:375)
	at org.rust.lang.core.resolve.NameResolutionKt.processPathResolveVariants(NameResolution.kt:350)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePath$result$1.invoke(RsPathReferenceImpl.kt:144)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePath$result$1.invoke(RsPathReferenceImpl.kt)
	at org.rust.lang.core.resolve.ProcessorsKt.collectPathResolveVariants(Processors.kt:101)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath(RsPathReferenceImpl.kt:143)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath$default(RsPathReferenceImpl.kt:142)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:131)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:129)
	at org.rust.lang.core.resolve.ref.RsResolveCache$resolveWithCaching$$inlined$run$lambda$1.compute(RsResolveCache.kt:99)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:42)
	at org.rust.lang.core.resolve.ref.RsResolveCache.resolveWithCaching(RsResolveCache.kt:99)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedCachedMultiResolve(RsPathReferenceImpl.kt:89)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedMultiResolve(RsPathReferenceImpl.kt:85)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.multiResolve(RsPathReferenceImpl.kt:81)
	at org.rust.lang.core.resolve.NamespaceKt.getNamespaces(Namespace.kt:81)
	at org.rust.ide.annotator.RsErrorAnnotatorKt$duplicatesByNamespace$1.invoke(RsErrorAnnotator.kt:1192)
	at org.rust.ide.annotator.RsErrorAnnotatorKt$duplicatesByNamespace$duplicates$4.invoke(RsErrorAnnotator.kt:1211)
	at org.rust.ide.annotator.RsErrorAnnotatorKt$duplicatesByNamespace$duplicates$4.invoke(RsErrorAnnotator.kt)
	at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:315)
	at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:303)
	at org.rust.ide.annotator.RsErrorAnnotatorKt.duplicatesByNamespace(RsErrorAnnotator.kt:1342)
	at org.rust.ide.annotator.RsErrorAnnotatorKt.checkDuplicates(RsErrorAnnotator.kt:1126)
	at org.rust.ide.annotator.RsErrorAnnotatorKt.checkDuplicates$default(RsErrorAnnotator.kt:1122)
	at org.rust.ide.annotator.RsErrorAnnotator$annotateInternal$visitor$1.visitModItem(RsErrorAnnotator.kt:71)
	at org.rust.lang.core.psi.impl.RsModItemImpl.accept(RsModItemImpl.java:27)
	at org.rust.lang.core.psi.impl.RsModItemImpl.accept(RsModItemImpl.java:31)
	at org.rust.ide.annotator.RsErrorAnnotator.annotateInternal(RsErrorAnnotator.kt:113)
	at com.intellij.ide.annotator.AnnotatorBase.annotate(AnnotatorBase.kt:21)
	at com.intellij.codeInsight.daemon.impl.DefaultHighlightVisitor.runAnnotators(DefaultHighlightVisitor.java:136)
	at com.intellij.codeInsight.daemon.impl.DefaultHighlightVisitor.visit(DefaultHighlightVisitor.java:116)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.runVisitors(GeneralHighlightingPass.java:336)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.lambda$collectHighlights$5(GeneralHighlightingPass.java:269)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.analyzeByVisitors(GeneralHighlightingPass.java:295)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.lambda$analyzeByVisitors$6(GeneralHighlightingPass.java:298)
	at com.intellij.codeInsight.daemon.impl.DefaultHighlightVisitor.analyze(DefaultHighlightVisitor.java:96)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.analyzeByVisitors(GeneralHighlightingPass.java:298)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.collectHighlights(GeneralHighlightingPass.java:266)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.collectInformationWithProgress(GeneralHighlightingPass.java:212)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:84)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:56)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$1(PassExecutorService.java:400)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1137)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:393)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:392)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:368)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:172)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:183)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:366)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:188)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

Reply:
Indeed, looks like there is bug related to `rustStructureModificationTracker`. Simple workaround would be to manually call `project.rustPsiManager.incRustStructureModificationCount()` after deleting `RsUseItem`. Proper solution probably requires changes to [`CacheInvalidator`](https://github.com/intellij-rust/intellij-rust/blob/029836c47f3c0710f6a40bf68ad0e44edf9dabb3/src/main/kotlin/org/rust/lang/core/psi/RsPsiManager.kt#L125-L143), e.g. add handling for `ChildRemoval.After` (cc @vlad20012)

Reply:
Fixed in #6431